[unsubscriptable-object] FP for generic classes (PEP 695)

### Bug description


The behavior described in [#9193](https://github.com/pylint-dev/pylint/issues/9193) is still observable with pylint 3.0.3 for classes.

```python
class Foo[T]:
    def __init__(self, value: T):
        self.value = value

x = Foo[int](1)
```

Works fine but `T` is marked as an undefined variable, and `Foo` is marked unsubscriptable.

### Configuration

_No response_

### Command used

```shell
pylint foo.py
```


### Pylint output

```python
************* Module foo
foo.py:1:0: C0114: Missing module docstring (missing-module-docstring)
foo.py:1:0: C0104: Disallowed name "foo" (disallowed-name)
foo.py:1:0: C0115: Missing class docstring (missing-class-docstring)
foo.py:2:30: E0602: Undefined variable 'T' (undefined-variable)
foo.py:1:0: R0903: Too few public methods (0/2) (too-few-public-methods)
foo.py:6:4: E1136: Value 'Foo' is unsubscriptable (unsubscriptable-object)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```


### Expected behavior

There should be no E0602 or E1136.

### Pylint version

```shell
pylint 3.0.3
astroid 3.0.2
Python 3.12.0 (main, Dec  4 2023, 20:32:31) [Clang 15.0.0 (clang-1500.0.40.1)]
```


### OS / Environment

macOS 14.2.1 / Apple M2 Max

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/9406.false_positive
+++ b/doc/whatsnew/fragments/9406.false_positive
@@ -1,0 +1,3 @@
+Fix multiple false positives for generic class syntax added in Python 3.12 (PEP 695).
+
+Closes #9406

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -41,7 +41,7 @@ dependencies    = [
     # Also upgrade requirements_test_min.txt.
     # Pinned to dev of second minor update to allow editable installs and fix primer issues,
     # see https://github.com/pylint-dev/astroid/issues/1341
-    "astroid>=3.2.1,<=3.3.0-dev0",
+    "astroid>=3.2.2,<=3.3.0-dev0",
     "isort>=4.2.5,<6,!=5.13.0",
     "mccabe>=0.6,<0.8",
     "tomli>=1.1.0;python_version<'3.11'",

--- a/requirements_test_min.txt
+++ b/requirements_test_min.txt
@@ -1,6 +1,6 @@
 .[testutils,spelling]
 # astroid dependency is also defined in pyproject.toml
-astroid==3.2.1  # Pinned to a specific version for tests
+astroid==3.2.2  # Pinned to a specific version for tests
 typing-extensions~=4.11
 py~=1.11.0
 pytest~=7.4

--- a/tests/functional/g/generic_class_syntax.py
+++ b/tests/functional/g/generic_class_syntax.py
@@ -1,0 +1,38 @@
+# pylint: disable=missing-docstring,too-few-public-methods
+from typing import Generic, TypeVar
+
+_T = TypeVar("_T")
+
+
+class Entity(Generic[_T]):
+    last_update: int | None = None
+
+    def __init__(self, data: _T) -> None:
+        self.data = data
+
+
+class Sensor(Entity[int]):
+    def __init__(self, data: int) -> None:
+        super().__init__(data)
+
+    def async_update(self) -> None:
+        self.data = 2
+
+        if self.last_update is None:
+            pass
+        self.last_update = 2
+
+
+class Switch(Entity[int]):
+    def __init__(self, data: int) -> None:
+        Entity.__init__(self, data)
+
+
+class Parent(Generic[_T]):
+    def __init__(self):
+        self.update_interval = 0
+
+
+class Child(Parent[_T]):
+    def func(self):
+        self.update_interval = None

--- a/tests/functional/g/generic_class_syntax.py
+++ b/tests/functional/g/generic_class_syntax.py
@@ -1,11 +1,11 @@
 # pylint: disable=missing-docstring,too-few-public-methods
-from typing import Generic, TypeVar
+from typing import Generic, TypeVar, Optional
 
 _T = TypeVar("_T")
 
 
 class Entity(Generic[_T]):
-    last_update: int | None = None
+    last_update: Optional[int] = None
 
     def __init__(self, data: _T) -> None:
         self.data = data

--- a/tests/functional/g/generic_class_syntax_py312.py
+++ b/tests/functional/g/generic_class_syntax_py312.py
@@ -1,0 +1,33 @@
+# pylint: disable=missing-docstring,too-few-public-methods
+class Entity[_T: float]:
+    last_update: int | None = None
+
+    def __init__(self, data: _T) -> None:  # [undefined-variable]  # false-positive
+        self.data = data
+
+
+class Sensor(Entity[int]):
+    def __init__(self, data: int) -> None:
+        super().__init__(data)
+
+    def async_update(self) -> None:
+        self.data = 2
+
+        if self.last_update is None:
+            pass
+        self.last_update = 2
+
+
+class Switch(Entity[int]):
+    def __init__(self, data: int) -> None:
+        Entity.__init__(self, data)
+
+
+class Parent[_T]:
+    def __init__(self):
+        self.update_interval = 0
+
+
+class Child[_T](Parent[_T]):  # [undefined-variable]  # false-positive
+    def func(self):
+        self.update_interval = None

--- a/tests/functional/g/generic_class_syntax_py312.rc
+++ b/tests/functional/g/generic_class_syntax_py312.rc
@@ -1,0 +1,2 @@
+[testoptions]
+min_pyver=3.12

--- a/tests/functional/g/generic_class_syntax_py312.txt
+++ b/tests/functional/g/generic_class_syntax_py312.txt
@@ -1,0 +1,2 @@
+undefined-variable:5:29:5:31:Entity.__init__:Undefined variable '_T':UNDEFINED
+undefined-variable:31:23:31:25:Child:Undefined variable '_T':UNDEFINED

